--- beaker/crypto/util.py	(buggy py3k version)
+++ beaker/crypto/util.py	(working copy)
@@ -15,16 +15,21 @@ except ImportError:
     # When using the stdlib, we have to make sure the hmac version and sha
     # version are compatible
     if util.py24:
-        from sha import sha as sha1
-        import sha as hmac_sha1
+        from sha import sha as _sha1
+        import sha as _hmac_sha1
     else:
         # NOTE: We have to use the callable with hashlib (hashlib.sha1),
         # otherwise hmac only accepts the sha module object itself
-        from hashlib import sha1
-        hmac_sha1 = sha1
+        from hashlib import sha1 as _sha1
+        _hmac_sha1 = _sha1
 
 
 if util.py24:
-    from md5 import md5
+    from md5 import md5 as _md5
 else:
-    from hashlib import md5
+    from hashlib import md5 as _md5
+
+# hackfix
+def md5(s): return _md5(bytes(s,'utf-8'))
+def sha1(s): return _sha1(bytes(s,'utf-8'))
+def hmac_sha1(s): return _hmac_sha1(bytes(s,'utf-8'))
